samples/src/Universal/Aspects/BaseUniversalWrapperAspect.cs

```diff
@@ -0,0 +1,133 @@
+            {
+                var syncResultType = eventArgs.ReturnType.IsConstructedGenericType ? eventArgs.ReturnType.GenericTypeArguments[0] : _voidTaskResult;
+
+                return _asyncHandler.MakeGenericMethod(syncResultType).Invoke(this, new object[] { target, eventArgs.Copy() });
+            }
+
+            var syncReturnType = eventArgs.ReturnType == typeof(void) ? typeof(object) : eventArgs.ReturnType;
+            return _syncHandler.MakeGenericMethod(syncReturnType).Invoke(this, new object[] { target, eventArgs.Copy() });
+        }
+
+        private T WrapSync<T>(Func<object[], object> target, AspectEventArgs eventArgs)
+        {
+            OnBefore(eventArgs);
+
+            try
+            {
+                var result = (T)target(eventArgs.Args);
+
+                OnAfter(eventArgs);
+
+                return result;
+            }
+            catch (Exception exception)
+            {
+                OnException(AspectExceptionEventArgs.CreateFrom(eventArgs, exception));
+
+                return default;
+            }
+        }
+
+        private async Task<T> WrapAsync<T>(Func<object[], object> target, AspectEventArgs eventArgs)
+        {
+            OnBefore(eventArgs);
+
+            try
+            {
+                var result = await (Task<T>)target(eventArgs.Args);
+
+                OnAfter(eventArgs);
+
+                return result;
+            }
+            catch (Exception exception)
+            {
+                OnException(AspectExceptionEventArgs.CreateFrom(eventArgs, exception));
+
+                return default;
+            }
+        }
+
+        private void OnBefore(AspectEventArgs eventArgs)
+        {
+            foreach (var attr in eventArgs.Triggers)
+            {
+                try
+                {
+                    attr.OnBefore(eventArgs);
+                }
+                catch
+                {
+                    // In case the overridden OnBefore throws an exception, ignore this exception so that the other OnBefore methods are still called from here.
+                }
+            }
+        }
+
+        private void OnAfter(AspectEventArgs eventArgs)
+        {
+            foreach (var attr in eventArgs.Triggers)
+            {
+                try
+                {
+                    attr.OnAfter(eventArgs);
+                }
+                catch
+                {
+                    // In case the overridden OnBefore throws an exception, ignore this exception so that the other OnBefore methods are still called from here.
+                }
+            }
+        }
+
+        private void OnException(AspectExceptionEventArgs eventArgs)
+        {
+            foreach (var attr in eventArgs.Triggers)
+            {
+                try
+                {
+                    attr.OnException(eventArgs);
+                }
+                catch
+                {
+                    // TODO : need to think about what we want to do here...

```

Comment:
@pamidur What do you think on this?

Reply:
What options we have?
1. Remove try catch - so that user might choose to throw an exception in OnBefore for some business logic related reasons
2. Keep try catch and add Debugging.Trace logging - this way we make sure all OnBefore methods have a chance to run but user don't have an option to terminate an execution based on business logic.
3. Introduce more complex flow where we execute all OnBefore then gather exceptions if any, and throw it as an AggregateException

I'm leaning to the first one. WDYT?

Reply:
With *Handle OnBefore and OnAfter with option 2* - user cannot have a scenario where user wants to check access rights before executing method and prevent an executions in case of violation

Reply:
In that case, option 1 sounds simple and logical. So I think we should use that one.

Reply:
we can go with option 1 for now, and might get back to option 3 later